fdbclient/ManagementAPI.actor.cpp

```diff
@@ -1587,6 +1589,69 @@ ACTOR Future<Void> excludeServers(Database cx, vector<AddressExclusion> servers,
 				wait(tr.commit());
 				return Void();
 			} catch (Error& e) {
+				TraceEvent("ExcludeServersError").error(e, true);
+				wait(tr.onError(e));
+			}
+		}
+	}
+}
+
+void excludeLocalities(Transaction& tr, std::unordered_set<std::string>& localities, bool failed) {
+	tr.setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
+	tr.setOption(FDBTransactionOptions::PRIORITY_SYSTEM_IMMEDIATE);
+	tr.setOption(FDBTransactionOptions::LOCK_AWARE);
+	tr.setOption(FDBTransactionOptions::USE_PROVISIONAL_PROXIES);
+	std::string excludeVersionKey = deterministicRandom()->randomUniqueID().toString();
+	auto localityVersionKey = failed ? failedLocalityVersionKey : excludedLocalityVersionKey;
+	tr.addReadConflictRange(singleKeyRange(localityVersionKey)); // To conflict with parallel includeLocalities
+	tr.set(localityVersionKey, excludeVersionKey);
+	for (auto& l : localities) {
+		if (failed) {
+			tr.set(encodeFailedLocalityKey(l), StringRef());
+		} else {
+			tr.set(encodeExcludedLocalityKey(l), StringRef());
+		}
+	}
+	TraceEvent("ExcludeLocalitiesCommit")
+		.detail("Localities", describe(localities))
+		.detail("ExcludeFailed", failed);
+}
+
+ACTOR Future<Void> excludeLocalities(Database cx, std::unordered_set<std::string> localities, bool failed) {
+	if (cx->apiVersionAtLeast(700)) {
+		state ReadYourWritesTransaction ryw(cx);
+		loop {
+			try {
+				ryw.setOption(FDBTransactionOptions::SPECIAL_KEY_SPACE_ENABLE_WRITES);
+				ryw.set(
+				    SpecialKeySpace::getManagementApiCommandOptionSpecialKey(failed ? "failedlocality" : "excludedlocality", "force"),
+				    ValueRef());
+				for (auto& l : localities) {
+					Key addr = failed
+					               ? SpecialKeySpace::getManagementApiCommandPrefix("failedlocality").withSuffix(l)
+					               : SpecialKeySpace::getManagementApiCommandPrefix("excludedlocality").withSuffix(l);
+					ryw.set(addr, ValueRef());
+				}
+				TraceEvent("ExcludeLocalitiesSpecialKeySpaceCommit")
+				    .detail("Localities", describe(localities))
+					.detail("ExcludeFailed", failed);
+
+				wait(ryw.commit());
+				return Void();
+			} catch (Error& e) {
+				TraceEvent("ExcludeLocalitiesError").error(e, true);
+				wait(ryw.onError(e));
+			}
+		}
+	} else {
```

Comment:
@apkar Is this urgent enough for 6.3? 

 backwards compatibility: exclude by locality is a new "feature". I'm unaware we have our backward compatibility guarantee for such change. IMHO, if we support this, say, in 6.3.x,  `exclude by locality` in version < 6.3.x won't be supported. 

`exclude failed by IP` is still supported.

Reply:
If we do decide to go down that road, it would be ideal if any attempt to exclude by locality would fail if the cluster doesn't support it. Absent that, it should at least be well understood what the behavior is when you do it as well as what happens if you then upgrade to a cluster that supports it. We should make sure it doesn't react in an undesirable way if some of these keys are set, etc.

Reply:
Sorry missed this until now. I would really like this in 6.3, as 7.0 seems bit far in future. I agree on the backward compatibility concerns, do we have checks to protect against a process/cluster going back on old patch releases?

Reply:
> as 7.0 seems bit far in future

For now this is going into 7.1 -- any earlier version will need some agreement. I am not a big fan of porting features back and it should be only done in very rare cases (whether this is one, I don't know). I think a PR is the wrong place to discuss this. But generally I believe one of the main reasons our release processes are so long is because we back port so many features